draft-ietf-quic-recovery.md

```diff
@@ -1122,15 +1123,21 @@ Pseudocode for OnLossDetectionTimeout follows:
 ~~~
 OnLossDetectionTimeout():
   loss_time, pn_space = GetEarliestLossTime()
-  if (loss_time != 0):
-    // Time threshold loss Detection
-    DetectLostPackets(pn_space)
   // Retransmit crypto data if no packets were lost
   // and there are still crypto packets in flight.
-  else if (crypto packets are in flight):
+  if (crypto packets are in flight):
     // Crypto retransmission timeout.
     RetransmitUnackedCryptoData()
     crypto_count++
+  else if (endpoint is client without 1-RTT keys):
+    if (will send Initial packet):
```

Comment:
I'd say
"if (has Handshake keys):
  SendOneHandshakePacket()
 else
  SendOnePaddedInitialPacket()"

Reply:
@martinthomson, is this consistent with your earlier suggestions?

Reply:
No. The important factor is not whether you have handshake keys, but whether your peer believes your address to be validated.

Reply:
To be clear, my suggestion started at line 1133.  The intent was to change "Will send Initial packet" to "has Handshake keys", which I believe is clearer.

Reply:
"Will send Initial packet" is correct.

Reply:
@martinthomson Why?  In particular, do you think it is more correct?

This is already in a conditional that indicate the 1RTT keys aren't present.

@Ralith please add a comment above this section as well, to explain why it's here, since it may not be obvious.

Reply:
I think I understand @martinthomson's point -- you want to keep sending initial packets until you know that the server has received and processed a handshake packet, which is when the server considers the client validated. This would happen when the client receives an ACK in a Handshake packet (for a Handshake packet) or has the 1-RTT keys.

That said, I agree with @ianswett that "will send Initial packet" is an odd predicate. I think we can say (has 1-RTT keys or has received an ACK in a Handshake packet). I could be wrong, but I think this captures the condition we're looking for.

Reply:
That sounds like the approach that I originally had here, which was [argued against](https://github.com/quicwg/base-drafts/pull/2281/files/8088a86a41852b071d3fee371357da4cf56e3a12#discussion_r248056667).

Reply:
Martin didn't want to allow for ACK of handshake stopping the anti-amplification retransmissions, which I think is fine, though a bit overly strict.  So all of this is within the "if (endpoint is client without 1-RTT keys)" conditional, which is our proxy for: Should we keep trying to prevent a deadlock?

To be clear on what I'm suggesting:

else if (endpoint is client without 1-RTT keys):
  // The client must continue sending packets until it's
  // sure the server has validated the path.
  if (has Handshake keys):
    SendOneHandshakePacket()
  else
    SendOnePaddedInitialPacket()

Reply:
I see. Your proposal seems good to me. @martinthomson: can you clarify what you think is missing?

Reply:
Striking out the bit that we've agreed to remove, the text from the transport draft is:

> Clients MUST ensure that UDP datagrams containing Initial packets are sized to at least 1200 bytes, adding padding to packets in the datagram as necessary. ~~Once a client has received an acknowledgment for a Handshake packet it MAY send smaller datagrams.~~ Sending padded datagrams ensures that the server is not overly constrained by the amplification restriction.

This contradicts that text.  It's true that a server that received a Handshake packet is no longer subject to the anti-DoS limit on bytes transmitted, but it's a bit weird that we have an algorithm that directly contradicts a MUST-level requirement in the other document.

Reply:
@martinthomson: AIUI, you'd remember this better since it was your PR, but wasn't the bit we agreed to remove was about special-casing padding of Initial packets? My reading of @ianswett 's proposed text doesn't contradict that.